### ISSUE TITLE
mpi4py: Deadlock in singleton OSC tests

### ISSUE THREAD

**Issue by dalcinl** [open-mpi/ompi#11250]
There is a new regression in mpi4py tests running in singleton MPI mode within GitHub Actions. The regression was likely introduced over the last 3/4 days, after commit c6fd8f4effa0f583c5d2c72eb5fbf66dc74c70a1. This time, there is a deadlock seems to happen whthin a test exercising RMA accumulate with `MPI_PROC_NULL` target and `MPI_SUM` operation [[logs]](https://github.com/mpi4py/mpi4py-testing/actions/runs/3769021714/jobs/6407899225).

**Comment by dalcinl** [comment]
@rhc54 Any chance this could be related to #11231 ?

**Comment by rhc54** [comment]
I don't see how - my change had nothing to do with MPI operations. I just fixed the comm_spawn operation so procs would be started.

**Comment by dalcinl** [comment]
> I don't see how

Sorry, it was just wild speculation. I can confirm things are OK up to #11231 [[logs]](https://github.com/mpi4py/mpi4py-testing/actions/runs/3771743258/jobs/6412298293).

**Comment by dalcinl** [comment]
@wckzhang @devreal The regression was introduced in #11214 [[logs]](https://github.com/mpi4py/mpi4py-testing/actions/runs/3771799211/jobs/6412380602).

**Comment by rhc54** [comment]
FWIW: there are hundreds of failures in OMPI main in MPI ops showing in MTT. Not sure when they started - they were present a couple of days ago when I looked, and still there last night.

The OMPI v5 branch failures are waiting on PR https://github.com/open-mpi/ompi/pull/11232 to be committed that will fix the majority of them.

**Comment by dalcinl** [comment]
@jsquyres Could you please help me get this issue addressed? mpi4py testing against ompi@master has been broken for two weeks, and if things don't get back to normal, new issues may go unnoticed and accumulate on top of each other. Last time such things happened, it took  months of triage and close watching to recover.

**Comment by jsquyres** [comment]
@open-mpi/ompi-rm-5-0-x This issue should probably get added to the v5.0 list.

**Comment by dalcinl** [comment]
> @open-mpi/ompi-rm-5-0-x This issue should probably get added to the v5.0 list.

I'm not sure who you are mentioning/notifying, but just in case, I clarify: This issue only happens with the main branch, not with v5.0.x [[see builds]](https://github.com/mpi4py/mpi4py-testing/actions/runs/3859574560)

**Comment by jsquyres** [comment]
@dalcinl Sorry for the delay on replying - I just got back in the office on Monday after vacation and am in catch-up mode.  I'm specifically mentioning the v5.0.x release managers.

So the issue I'm raising is: main and v5.0.x *should* be in pretty close sync at this point.  The fact that it happens on main and does *not* happen on v5.0.x is also somewhat concerning.  I.e., are there other fixes on main that did not make it over to v5.0.x?  (even if there's still another fix that is needed on main)

**Comment by awlauria** [comment]
Good questions.

We haven't done a full sweep of what isn't on v5.0.0 in a while - we've just been taking things as they came over. So it isn't surprising that there is a bug on main that isn't on v5 (yet). That's sort of by design, we've stopped the large PR syncs to try and not disrupt things as we get closer.

Just eyeballing osc/rdma, I see a small handful of commits that might have caused the regression, just based on the date reported:
202b50f40b3dfca40cc59f57321fb69fea2cbfae
d3d661c3613776cb2ddfb2be081d307026d726f4

**Comment by dalcinl** [comment]
@awlauria As I commented [above](https://github.com/open-mpi/ompi/issues/11250#issuecomment-1364544072), the regression was introduced in #11214 [[logs]](https://github.com/mpi4py/mpi4py-testing/actions/runs/3771799211/jobs/6412380602).

**Comment by awlauria** [comment]
Ah - thanks. @wckzhang would you be able to take a look?

**Comment by wckzhang** [comment]
I will take a look, if it's not happening in v5.0.x it's because I haven't backported it there yet. Let me go dive deep into this one.

**Comment by wckzhang** [comment]
@dalcinl Is there a command I can use to reproduce locally?

Looking at the repo is it something like: ```mpiexec -n 1 python mpi4py/test/runtests.py -v```

**Comment by dalcinl** [comment]
@wckzhang  Actually, the hang occurs when using singleton MPI init mode, so you sould invoke without mpiexec.

To run the whole testsuite:
```
 python test/runtests.py -v
```
If you want to narrow down to the minimum to trigger the issue quickier, do
```
 python test/test_rma.py -v -k ProcNull
```

**Comment by wckzhang** [comment]
Thanks, I will try to reproduce it today

**Comment by wckzhang** [comment]
@dalcinl It seems to be passing when I run that on main: branch of OMPI

```
[ec2-user@c5n-dy-c5n18xlarge-1 mpi4py]$ python3 test/test_rma.py -v -k ProcNull
[0@c5n-dy-c5n18xlarge-1] Python 3.7.15 (/usr/bin/python3)
[0@c5n-dy-c5n18xlarge-1] numpy 1.21.6 (/home/ec2-user/.local/lib/python3.7/site-packages/numpy)
[0@c5n-dy-c5n18xlarge-1] MPI 3.1 (Open MPI 5.1.0)
[0@c5n-dy-c5n18xlarge-1] mpi4py 4.0.0.dev0 (/home/ec2-user/.local/lib/python3.7/site-packages/mpi4py)
testAccumulateProcNullReplace (__main__.TestRMASelf) ... ok
testAccumulateProcNullSum (__main__.TestRMASelf) ... ok
testGetAccumulateProcNull (__main__.TestRMASelf) ... ok
testGetProcNull (__main__.TestRMASelf) ... ok
testPutProcNull (__main__.TestRMASelf) ... ok
testAccumulateProcNullReplace (__main__.TestRMAWorld) ... ok
testAccumulateProcNullSum (__main__.TestRMAWorld) ... ok
testGetAccumulateProcNull (__main__.TestRMAWorld) ... ok
testGetProcNull (__main__.TestRMAWorld) ... ok
testPutProcNull (__main__.TestRMAWorld) ... ok

----------------------------------------------------------------------
Ran 10 tests in 0.053s

OK
```

Do you have any advice?

**Comment by wckzhang** [comment]
I wasn't able to reproduce this issue but I did find another issue in the cuda delayed init code which I'll make a separate issue for

**Comment by dalcinl** [comment]
@wckzhang How are you configuring Open MPI? 

I'm using the following configure flags on a Fedora 37 system:
```sh
options=(
    --prefix=/home/devel/mpi/openmpi/dev
    --without-ofi
    --without-ucx
    --without-psm2
    --with-pmix=internal
    --with-prrte=internal
    --with-libevent=internal
    --with-hwloc=internal
    --enable-debug
    --enable-mem-debug
    --disable-man-pages
    --disable-sphinx
)
./configure "${options[@]}"
```

The following sequence of commands hangs:
```
$ git clone https://github.com/mpi4py/mpi4py.git
$ cd mpi4py
$ make # to build locally
$ python test/runtests.py -v -i test_rma
[0@kw61149] Python 3.11.1 (/usr/bin/python)
[0@kw61149] numpy 1.22.0 (/usr/lib64/python3.11/site-packages/numpy)
[0@kw61149] MPI 3.1 (Open MPI 5.1.0)
[0@kw61149] mpi4py 4.0.0.dev0 (/home/dalcinl/Devel/mpi4py/build/lib.linux-x86_64-cpython-311/mpi4py)
testAccumulate (test_rma.TestRMASelf.testAccumulate) ... ok
testAccumulateProcNullReplace (test_rma.TestRMASelf.testAccumulateProcNullReplace) ... ok
testAccumulateProcNullSum (test_rma.TestRMASelf.testAccumulateProcNullSum) ... ok
testCompareAndSwap (test_rma.TestRMASelf.testCompareAndSwap) ... 
  testCompareAndSwap (test_rma.TestRMASelf.testCompareAndSwap) (disp=0, rank=0, typecode='h', backend='array') ... ERROR
```

**Comment by lrbison** [comment]
@dalcinl yes, I can reproduce it now.  Thank you.

Seems that ompi_osc_rdma_lock_atomic_internal is stuck in deadlock, successfully getting the global lock but failing to acquire the peer's lock.

I'm looking into it now, thank you.

**Comment by lrbison** [comment]
I'm still debugging the root cause of the CAS failure, but the deadlock is actually the test harness not calling WIN.Unlock after an exception is raised, and then then the second test is not able to execute.

**Comment by lrbison** [comment]
The root cause should be fixed by https://github.com/open-mpi/ompi/pull/11304

OSC RDMA compare-and-swap had a path which returned an uninitialized value instead of success.

Thank you for the report @dalcinl !
